src/compiler/checker.ts

```diff
@@ -17029,31 +17026,22 @@ namespace ts {
             return false;
         }
 
-        // Return true if type is of the given kind. A union type is of a given kind if all constituent types
-        // are of the given kind. An intersection type is of a given kind if at least one constituent type is
-        // of the given kind.
-        function isTypeOfKind(type: Type, kind: TypeFlags): boolean {
-            if (type.flags & kind) {
+        function isTypeAssignableToKind(source: Type, kind: TypeFlags, strict?: boolean): boolean {
+            if (source.flags & kind) {
                 return true;
             }
-            if (type.flags & TypeFlags.Union) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (!isTypeOfKind(t, kind)) {
-                        return false;
-                    }
-                }
-                return true;
-            }
-            if (type.flags & TypeFlags.Intersection) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (isTypeOfKind(t, kind)) {
-                        return true;
-                    }
-                }
+            if (strict && source.flags & (TypeFlags.Any | TypeFlags.Void | TypeFlags.Undefined | TypeFlags.Null)) {
+                return false;
             }
-            return false;
+            return kind & TypeFlags.NumberLike && isTypeAssignableTo(source, numberType) ||

```

Comment:
Calling `isTypeAssignableTo` so many times seems fairly expensive if `source` is not a simple type. While the assignability check has a fast path for simple types via `isSimpleTypeRelatedTo`, if the relation doesn't hold we fall back to the more complex `checkTypeRelatedTo` function. I would recommend you determine if there is any additional check time overhead with this approach.